### DISCUSSION TITLE
Question about styled-components

### DISCUSSION THREAD

**Discussion by flavio-ever** [ben-rogerson/twin.macro#571] [Q&A] [answered]
Hello good afternoon. Sorry for my English.

I didn't find it in the documentation, and in some examples I see out there, they end up resulting in a syntax or typing error. I'm experimenting on nextjs, everything is set up fine (I think).

**Example:**
If I use "css" in the element, it triggers a typing error, pointing out that it doesn't exist and so on. I saw mentions of changing the "jsxImportSource" in the typescript, but as I'm using styled-components, I wasn't successful.

**Example:**
Is it okay to do it this way?

```js
import tw from 'twin.macro'
import styled from 'styled-components'

export const App = styled.div`
  ${tw`flex justify-center items-center h-screen`}
`
```

I see many articles about importing "styled" straight from the "twin.macro" module but apparently I didn't find anything inside. Maybe they've removed it, or need something else I don't know yet.

**Comment by ben-rogerson** [comment, marked as answer]
You can import styled but may run into some issues - the safe way is to import `styled` through twin.

Here's how I'd change your example:

```js
import tw, { styled } from 'twin.macro'

export const App = tw.div`flex justify-center items-center h-screen`
```

The styled import works great in the following typescript twin examples, could you let me know where you saw the broken implementation of styled?

- [snowpack-react-styled-components-typescript](https://github.com/ben-rogerson/twin.examples/tree/master/snowpack-react-styled-components-typescript)
- [storybook-styled-components-typescript](https://github.com/ben-rogerson/twin.examples/tree/master/storybook-styled-components-typescript)
- [vite-styled-components-typescript](https://github.com/ben-rogerson/twin.examples/tree/master/vite-styled-components-typescript)
- [webpack-styled-components-typescript](https://github.com/ben-rogerson/twin.examples/tree/master/webpack-styled-components-typescript)

**Reply by flavio-ever** [reply]
Hi, thanks for the feedback. So I won't use "styled-components" imports anymore

**I am using Typescript**

I tested the component examples by:
https://github.com/ben-rogerson/twin.macro

I tried the settings by:
https://github.com/ben-rogerson/twin.examples/tree/master/next-styled-components

```js
import tw, { styled } from "twin.macro"

const StyledInput = styled.input`
  color: black;
  ${({ hasBorder }) => hasBorder && tw`border-purple-500`}
const Input = () => <StyledInput hasBorder />
``` 

Import: 
`Module '"twin.macro"' has no exported member 'styled'. Did you mean to use 'import styled from "twin.macro"' instead?ts(2614)`

---

```js
import tw, { css, styled } from "twin.macro"

const Input = ({ hasHover }) => (
  <input css={[tw`border`, hasHover && tw`hover:border-black`]} />
)
``` 

Import:
`Module '"twin.macro"' has no exported member 'css'. Did you mean to use 'import css from "twin.macro"' instead?ts(2614)`

Element: 
`Type '{ children: Element; css: TwStyle[]; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.
  Property 'css' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.ts(2322)
(JSX attribute) css: TwStyle[]
`

**Reply by flavio-ever** [reply]
I found this year 2020 pull

https://github.com/ben-rogerson/twin.macro/pull/78

**Reply by flavio-ever** [reply]
I found this year 2020 issue
https://github.com/ben-rogerson/twin.macro/issues/77

**Reply by flavio-ever** [reply]
I ended up defining this paleative solution `.\@types\twin.d.ts` 

```js
import 'twin.macro'
import styled, { css } from 'styled-components'
declare module 'twin.macro' {
  export { css, styled }
}
```

Any predictions to fix or update the documentation?

**Reply by ben-rogerson** [reply]
Check the `types/twin.d.ts` files in those repos I posted above for the full set of types.

Here's the contents from one of them:

```js
// types/twin.d.ts
import 'twin.macro'
import styledImport, { CSSProp, css as cssImport } from 'styled-components'

declare module 'twin.macro' {
  // The styled and css imports
  const styled: typeof styledImport
  const css: typeof cssImport
}

declare module 'react' {
  // The css prop
  interface HTMLAttributes<T> extends DOMAttributes<T> {
    css?: CSSProp
  }
  // The inline svg css prop
  interface SVGProps<T> extends SVGProps<SVGSVGElement> {
    css?: CSSProp
  }
}

// The 'as' prop on styled components
declare global {
  namespace JSX {
    interface IntrinsicAttributes<T> extends DOMAttributes<T> {
      as?: string
    }
  }
}
```

You're right about the documentation on those styled-components examples - they definitely need some work!

They need new readme's, added in the same format as the [react-emotion-typescript](https://github.com/ben-rogerson/twin.examples/tree/master/react-emotion-typescript) readme. 
I understand this is a large job, but I'd really appreciate even one of the examples being updated if you have any capacity 👍

**Reply by flavio-ever** [reply]
I thank you for your feedback, I do hope to be able to contribute in the future, even if a little. I really like the work!

**Comment by trepechov-lime** [comment]
I've recently hit some issue with the types/twin.d.ts

I'm using headless ui so I had to extend the 
```
interface IntrinsicAttributes<T> extends DOMAttributes<T> {
    as?: string | Element | ExoticComponent
}

```
to support **as={Fragment}** but it introduces ts error on all Headless ui components (<Menu.Button>, <Dialog.Title> etc) with **as** property
